server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -801,15 +770,50 @@ private long relativeTime() {
         return relativeTimeProvider.getAsLong();
     }
 
-    private void processBulkIndexIngestRequest(
+    private boolean preprocessBulkRequest(Task task, BulkRequest bulkRequest, String executorName, ActionListener<BulkResponse> listener) {
+        final Metadata metadata = clusterService.state().getMetadata();
+        final Version minNodeVersion = clusterService.state().getNodes().getMinNodeVersion();
+        boolean needsProcessing = false;
+        for (BulkRequestPreprocessor preprocessor : bulkRequestPreprocessors) {
```

Comment:
There are now two preprocessors for requests:
- Ingest pipelines
- Field inference

they will run in order for updating the requests

Reply:
We need to be careful of the order here and make sure `ingest` ALWAYS runs first as "field inference" is the last step before indexing.

Reply:
We could ensure that with an assertion that checks it 👍